### PR TITLE
osdocs-415 adding workers to user-provisioned (AWS)

### PR DESCRIPTION
**Pull request by kalexand-rh**
https://jira.coreos.com/browse/OSDOCS-415

### CODE REVIEW

**Comment by kalexand-rh** [comment]
@abhinavdahiya, @sosiouxme, @jstuever, will you PTAL at these draft topics about adding more compute machines to UPI clusters? 

For AWS: https://osdocs415--ocpdocs.netlify.com/openshift-enterprise/latest/machine_management/user_infra/adding-aws-compute-user-infra.html
For vSphere: https://osdocs415--ocpdocs.netlify.com/openshift-enterprise/latest/machine_management/user_infra/adding-vsphere-compute-user-infra.html
For bare metal: https://osdocs415--ocpdocs.netlify.com/openshift-enterprise/latest/machine_management/user_infra/adding-bare-metal-compute-user-infra.html

File: modules/machine-vsphere-machines.adoc
Comment (DanyC97) [inline modules/machine-vsphere-machines.adoc]: fyi i've noticed that the guestinfo params are not taken into account following the above steps, however in the next tab - `vApp_Properties` the params are exposed and so the OP only need to fill in the data value.

If you look in the ova template you will see the properties

```

Properties:
  Key:         guestinfo.ignition.config.data
  Label:       Ignition config data
  Type:        string
  Description: Inline Ignition config data

  Key:         guestinfo.ignition.config.data.encoding
  Label:       Ignition config data encoding
  Type:        string
  Description: Encoding for Ignition config data

```
@staebler fyi ^^

**Comment by kalexand-rh** [comment]
@abhinavdahiya, @sosiouxme, @jstuever, will you PTAL at these draft topics about adding more compute machines to UPI clusters?

**Comment by kalexand-rh** [comment]
@abhinavdahiya, @sosiouxme, @jstuever, will you PTAL at these draft topics about adding more compute machines to UPI clusters?

**Comment by kalexand-rh** [comment]
@jianlinliu, can you help me get a review for this PR?

File: machine_management/user_infra/adding-aws-compute-user-infra.adoc
Comment (jianlinliu) [inline machine_management/user_infra/adding-aws-compute-user-infra.adoc]: Per my testing, after user launched a new worker machine, the machine can not listed in the ouptut of `oc get node`, unless approve the pending clieNt request CSR.
`
[root@preserve-jialiu-ansible ~]# oc get nodes
NAME                                        STATUS   ROLES    AGE    VERSION
ip-10-0-48-181.us-east-2.compute.internal   Ready    master   134m   v1.14.6+463c73f1f
ip-10-0-54-126.us-east-2.compute.internal   Ready    worker   126m   v1.14.6+463c73f1f
ip-10-0-57-129.us-east-2.compute.internal   Ready    master   134m   v1.14.6+463c73f1f
ip-10-0-65-35.us-east-2.compute.internal    Ready    worker   126m   v1.14.6+463c73f1f
ip-10-0-79-1.us-east-2.compute.internal     Ready    master   134m   v1.14.6+463c73f1f

[root@preserve-jialiu-ansible ~]# oc get csr
NAME        AGE     REQUESTOR                                                                   CONDITION
csr-dtwf2   8m26s   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending

[root@preserve-jialiu-ansible ~]# oc get csr -ojson | jq -r '.items[] | select(.status == {} ) | .metadata.name' | xargs oc adm certificate approve
certificatesigningrequest.certificates.k8s.io/csr-dtwf2 approved

[root@preserve-jialiu-ansible ~]# oc get node
NAME                                        STATUS     ROLES    AGE    VERSION
ip-10-0-48-181.us-east-2.compute.internal   Ready      master   135m   v1.14.6+463c73f1f
ip-10-0-54-126.us-east-2.compute.internal   Ready      worker   127m   v1.14.6+463c73f1f
ip-10-0-57-129.us-east-2.compute.internal   Ready      master   135m   v1.14.6+463c73f1f
ip-10-0-65-35.us-east-2.compute.internal    Ready      worker   127m   v1.14.6+463c73f1f
ip-10-0-77-219.us-east-2.compute.internal   NotReady   worker   5s     v1.14.6+463c73f1f
ip-10-0-79-1.us-east-2.compute.internal     Ready      master   135m   v1.14.6+463c73f1f

[root@preserve-jialiu-ansible ~]# oc get csr
NAME        AGE     REQUESTOR                                                                   CONDITION
csr-dtwf2   9m26s   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-xbnvf   7s      system:node:ip-10-0-77-219.us-east-2.compute.internal                       Pending

[root@preserve-jialiu-ansible ~]# oc get csr -ojson | jq -r '.items[] | select(.status == {} ) | .metadata.name' | xargs oc adm certificate approve
certificatesigningrequest.certificates.k8s.io/csr-xbnvf approved

[root@preserve-jialiu-ansible ~]# oc get csr
NAME        AGE   REQUESTOR                                                                   CONDITION
csr-dtwf2   10m   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-xbnvf   64s   system:node:ip-10-0-77-219.us-east-2.compute.internal                       Approved,Issued

[root@preserve-jialiu-ansible ~]# oc get node
NAME                                        STATUS   ROLES    AGE    VERSION
ip-10-0-48-181.us-east-2.compute.internal   Ready    master   136m   v1.14.6+463c73f1f
ip-10-0-54-126.us-east-2.compute.internal   Ready    worker   128m   v1.14.6+463c73f1f
ip-10-0-57-129.us-east-2.compute.internal   Ready    master   136m   v1.14.6+463c73f1f
ip-10-0-65-35.us-east-2.compute.internal    Ready    worker   128m   v1.14.6+463c73f1f
ip-10-0-77-219.us-east-2.compute.internal   Ready    worker   71s    v1.14.6+463c73f1f
ip-10-0-79-1.us-east-2.compute.internal     Ready    master   136m   v1.14.6+463c73f1f
`

So I think the description in `Approving the CSRs for your machines` section need a bit adjust.

Comment (jstuever) [reply machine_management/user_infra/adding-aws-compute-user-infra.adoc]: That is correct, the node will not join the cluster until it's `system:serviceaccount` certificate is approved. Furthermore, I don't believe it will create the `system:node` certificate until after the node has joined the cluster.

**Review by jstuever — APPROVED** [review]
The AWS-UPI and RHEL sections look good to me. As mentioned by others, and out of scope of this PR, the CSR approval file could use some modifications. In addition, it should probably mention how to verify the CSR(s) belong to the nodes being added. The system:node CSRs include the node names directly in the `oc get csr` output. However, the system.serviceaccount CSRs have the node name hidden in the output of `oc describe csr`.

File: machine_management/user_infra/adding-aws-compute-user-infra.adoc
Comment (kalexand-rh) [reply machine_management/user_infra/adding-aws-compute-user-infra.adoc]: @jstuever, 
>  However, the `system.serviceaccount` CSRs have the node name hidden in the output of oc describe csr. 
Is the method that Jianlin pointed out the best way, or is there another step that I should add to clarify which certificate you're approving?

Comment (jstuever) [reply machine_management/user_infra/adding-aws-compute-user-infra.adoc]: It gets kinda messy when you are validating `system.serviceaccount` certificates. The node name is not stored directly in the CSR; instead it is part of the certificate itself. You have to describe each of them using `oc describe csr <CSR>` and look for the `Subject.Common Name` to match `system:node:<node_name>`

```
oc get csr
NAME        AGE     REQUESTOR                                                                   CONDITION
csr-6s9jv   10m     system:node:jstuev-vj98j-m-1.c.openshift-dev-installer.internal             Approved,Issued
csr-crvjt   10m     system:node:jstuev-vj98j-m-2.c.openshift-dev-installer.internal             Approved,Issued
csr-fwlbl   10m     system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-hscmt   10m     system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-kqpvl   3m54s   system:node:jstuev-vj98j-w-a-l2dz6.c.openshift-dev-installer.internal       Approved,Issued
csr-n565v   10m     system:node:jstuev-vj98j-m-0.c.openshift-dev-installer.internal             Approved,Issued
csr-n8dzm   10m     system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-p8rgm   3m53s   system:node:jstuev-vj98j-w-b-v8sq9.c.openshift-dev-installer.internal       Approved,Issued
csr-pfgpg   3m46s   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-sbdxs   3m32s   system:node:jstuev-vj98j-w-c-4cl5k.c.openshift-dev-installer.internal       Approved,Issued
csr-v84gt   4m7s    system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-z8cfs   4m8s    system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued

oc describe csr csr-z8cfs
Name:               csr-z8cfs
Labels:             <none>
Annotations:        <none>
CreationTimestamp:  Tue, 15 Oct 2019 12:35:56 -0700
Requesting User:    system:serviceaccount:openshift-machine-config-operator:node-bootstrapper
Status:             Approved,Issued
Subject:
         Common Name:    system:node:jstuev-vj98j-w-a-l2dz6.c.openshift-dev-installer.internal
         Serial Number:  
         Organization:   system:nodes
Events:  <none>
```

Comment (jstuever) [reply machine_management/user_infra/adding-aws-compute-user-infra.adoc]: The method Jianlin works just fine if they don't care to validate the certificates before they approve them.

File: modules/machine-user-infra-machines-iso.adoc
Comment (morenod) [inline modules/machine-user-infra-machines-iso.adoc]: change {op-sytem} to {op-system}

File: modules/machine-user-infra-machines-iso.adoc
Comment (morenod) [inline modules/machine-user-infra-machines-iso.adoc]: Change config.ign to worker.ign, this is the name used by openshift-install to generate file

File: modules/machine-user-infra-machines-pxe.adoc
Comment (morenod) [inline modules/machine-user-infra-machines-pxe.adoc]: change bootstrap.ign to worker.ign

File: modules/machine-user-infra-machines-pxe.adoc
Comment (morenod) [inline modules/machine-user-infra-machines-pxe.adoc]: We need the worker.ign file, not the bootstrap one. change "location of the bootstrap Ignition config file" to "location of the worker Ignition config file"

File: modules/machine-user-infra-machines-pxe.adoc
Comment (morenod) [inline modules/machine-user-infra-machines-pxe.adoc]: same, worker.ign

File: modules/machine-user-infra-machines-pxe.adoc
Comment (morenod) [inline modules/machine-user-infra-machines-pxe.adoc]: same, worker instead of bootstrap

**Review by morenod — CHANGES_REQUESTED** [review]

File: machine_management/user_infra/adding-bare-metal-compute-user-infra.adoc
Comment (morenod) [reply machine_management/user_infra/adding-bare-metal-compute-user-infra.adoc]: Sorry, I didnt see the notification for this. I have added some comments.

@kalexand-rh as a general comment for all the files, we are using "compute" and "worker" to define the same set of servers. To avoid understanding problems, we should use only one of them. I suggest to use "worker" because compute is already being used to define Openstack servers

Comment (kalexand-rh) [reply machine_management/user_infra/adding-bare-metal-compute-user-infra.adoc]: @morenod, the install team's said since before 4.1 GA that they prefer calling the "workers" the compute machines.

Comment (morenod) [reply machine_management/user_infra/adding-bare-metal-compute-user-infra.adoc]: Ok, then we should remove all "workers" references or poin somewhere that workers and computes are the same

File: modules/machine-user-infra-machines-iso.adoc
Comment (kalexand-rh) [reply modules/machine-user-infra-machines-iso.adoc]: @morenod, is this true only when you add more compute/worker machines, or is it also true during cluster initialization?

File: modules/machine-user-infra-machines-pxe.adoc
Comment (kalexand-rh) [reply modules/machine-user-infra-machines-pxe.adoc]: Just when you add more workers, or also during initialization?

Comment (kalexand-rh) [reply modules/machine-user-infra-machines-pxe.adoc]: For all of these similar references, too. This method's reused from the installation flow, and I don't want to wreck that process.

Comment (morenod) [reply modules/machine-user-infra-machines-pxe.adoc]: When you are running worker installation, both initial install or scaling up, you need to use worker.ign file.

bootstrap.ign is only used for the bootstrap node, in the initial installation

File: modules/machine-user-infra-machines-iso.adoc
Comment (morenod) [reply modules/machine-user-infra-machines-iso.adoc]: always you install a worker, you need the worker.ign file

**Comment by ahardin-rh** [comment]
@kalexand-rh Any update on this? I think once we get this merged,we can close out https://bugzilla.redhat.com/show_bug.cgi?id=1753034  Thanks!

**Comment by codyhoag** [comment]
The vSphere and bare metal parts were extracted off of this PR and re-reviewed/merged in #25046. This PR now holds the AWS portion of this work.

**Comment by codyhoag** [comment]
@kalexand-rh I've resubmitted this in #30360. You've done all the heavy lifting for this one, so I'll try to push this through reviews so we can get it merged. I had gotten the bare metal and vSphere ones finished a while back (#25046), but had this sticking around because it wasn't as urgent for AWS. Hopefully a quick and painless review will get this in soon. :slightly_smiling_face:

**Comment by codyhoag** [comment]
All content from this PR has been extracted, approved, and merged:
* vSphere and bare metal: #25046
* AWS: #30360

**Pull request closed without merging**
